### PR TITLE
Complete flapping module

### PR DESCRIPTION
**Pull request by Gilbert-Wong**
NOTE: This is still a work-in-progress PR.

### CODE REVIEW

File: include/emqx.hrl [hunk 1]
```diff
@@ -161,6 +161,20 @@
...
           desc   :: binary(),
           until  :: integer()
-        }).
+       }).
+
+%%--------------------------------------------------------------------
+%% Flapping
+%%--------------------------------------------------------------------
+
+-record(flapping, {
```
Comment (spring2maz) [inline include/emqx.hrl]: why this record is put to a hrl?
it seems to be only used in emqx_flapping module.

File: include/emqx.hrl [hunk 2]
```diff
@@ -161,6 +161,20 @@
...
+          state         :: any(),
+          check_times   :: pos_integer(),
+          time_interval :: pos_integer(),
+          high_treshold :: float(),
```
Comment (spring2maz) [inline include/emqx.hrl]: threshold.

File: include/emqx.hrl (hunk 1 above)
Comment (spring2maz) [reply include/emqx.hrl]: you may have more to add to this PR, but from my superficial understanding of the current code, it's better to define it as an opaque structure for emqx_flapping

File: src/emqx_flapping.erl [hunk 3]
```diff
@@ -17,57 +17,57 @@
...
+        _ ->
+            FlappingRecord
+    end.
+
+-spec check_state_transition(non_neg_integer(), pos_integer(), pos_integer(), Fun::fun(), term(), number()) -> pos_integer().
+check_state_transition(0, _TimeInterval, _Total, _CheckState, _State, Weight) ->
+    Weight;
+check_state_transition(Count, TimeInterval, Total, CheckState, State, Weight)
+  when Count > 0->
+    timer:sleep(TimeInterval),
```
Comment (spring2maz) [inline src/emqx_flapping.erl]: this is scary, planing to run this is a swpan_link ?

**Comment by spring2maz** [comment]
I'd recommend to start this PR by documenting (as code comments) what is to achieve.
It would be helpful to share the links references to weighted state transition algorithm (including the reasoning behind the hard-coded magic float-point numbers)

Also sharing some of my thoughts:
* flapping detection is better to be done in a separate process
* periodic state check can miss some state changes while sleeping, a better approach is to send state change messages to checker.
* if periodic check is good enough, it is better to be triggered in a send_after message instead of timer:sleep
* there seem to lack of a callback for what to do when flapping is detected.

**Comment by Gilbert-Wong** [comment]
Thanks for your advice @spring2maz very much 👍 . The flapping module is still draft. I would take your advice into consideration seriously.

**Comment by Gilbert-Wong** [comment]
This branch was git branch from testing branch and is too old to have conflicts.

**Pull request closed without merging**